File: numerical-analysis/python/sparse.py
from collections import defaultdict
from fillMatrix import fillMatrix
import time
import csv


class SparseMatrix:

    def __init__(self, rows, cols, S):
        self.rows = rows
        self.cols = cols
        self.S = S

    @staticmethod
    def to_sparse(M):
        S = dict()
        for i, row in enumerate(M):
            for j, val in enumerate(row):
                if val:
                    S[(i, j)] = val
        return S

    @classmethod
    def from_dense(cls, M):
        rows, cols = len(M), len(M[0])
        S = cls.to_sparse(M)
        return cls(rows, cols, S)

    def multiply(self, B):
        A = self.S
        B = B.S
        C = defaultdict(int)
        for (a_r, a_c), a_val in A.items():
            for (b_r, b_c), b_val in B.items():
                if a_c == b_r:
                    C[(a_r, b_c)] += a_val * b_val
        return C


def recordTimes(filename, functionName):
    directory = "../matrices/" + filename
    matrix = fillMatrix(directory)
    matrix = SparseMatrix.from_dense(matrix)
    with open("../benchmarks.csv", "a") as f:
        writer = csv.writer(f)
        for _ in range(5):
            start = time.time()
            matrix.multiply(matrix)
            end = time.time()
            writer.writerow([
                "python",
                functionName,
                filename,
                matrix.rows,
                (end - start) * 1000
            ])


if __name__ == '__main__':

    recordTimes("Hamrle1.mtx", "sparse")
    recordTimes("GD99_b.mtx", "sparse")
    recordTimes("can_256.mtx", "sparse")
    recordTimes("dwa512.mtx", "sparse")
    recordTimes("delaunay_n10.mtx", "sparse")

File: numerical-analysis/cpp/sparse.cpp
#include <cstdio>
#include <chrono>
#include <vector>
#include <bits/stdc++.h>
#include <iostream>
#include <fstream>
#include "readFile.hpp"

using namespace std;
using namespace std::chrono;

struct hash_pair
{
    template <class T1, class T2>
    size_t operator()(const pair<T1, T2> &p) const
    {
        auto hash1 = hash<T1>{}(p.first);
        auto hash2 = hash<T2>{}(p.second);
        return hash1 ^ hash2;
    }
};

#define sparse_matrix unordered_map<pair<int, int>, double, hash_pair>

sparse_matrix to_sparse(vector<vector<double>> &matrix)
{
    sparse_matrix S;
    for (int i = 0; i < matrix.size(); ++i)
    {
        for (int j = 0; j < matrix[i].size(); ++j)
        {
            if (matrix[i][j])
                S[make_pair(i, j)] = matrix[i][j];
        }
    }
    return S;
}

sparse_matrix multiply(sparse_matrix &A, sparse_matrix &B)
{
    sparse_matrix C;
    for (auto &keyA : A)
    {
        for (auto &keyB : B)
        {
            if (keyA.first.second == keyB.first.first)
            {
                int x = (int)keyA.first.first, y = (int)keyB.first.second;
                C[make_pair(x, y)] += keyA.second * keyB.second;
            }
        }
    }
    return C;
}

void recordTimes(string filename)
{
    string directory = "../matrices/" + filename;
    std::ofstream myfile;
    myfile.open("../benchmarks.csv", std::ios_base::app); // append instead of overwrite
    vector<vector<double>> matrix = fill_matrix(directory);
    vector<vector<double>> res(matrix.size(), vector<double>(matrix[0].size(), 0.));

    sparse_matrix sparse = to_sparse(matrix);
    for (int i = 0; i < 5; ++i)
    {
        auto start = high_resolution_clock::now();
        multiply(sparse, sparse);
        auto stop = high_resolution_clock::now();
        auto duration = duration_cast<milliseconds>(stop - start);
        myfile << "cpp,sparse," << filename << "," << res.size() << "," << duration.count() << "\n";
    }
    myfile.close();
}

int main()
{
    recordTimes("Hamrle1.mtx");
    recordTimes("GD99_b.mtx");
    recordTimes("can_256.mtx");
    recordTimes("dwa512.mtx");
    recordTimes("delaunay_n10.mtx");
}

//! g++ sparse.cpp readFile.cpp -o out
// https://www.youtube.com/watch?v=3cdeehULUBo